Added model distillation guide

MichelBartels opened the pull request:
This adds a guide describing how and why to use model distillation. It also contains a table listing the performance improvements of model distillation for different models.

MichelBartels committed:
Add model distillation guide

MichelBartels committed:
Added explanation of train_filename and data_dir

MichelBartels committed:
Updated table with expected performance

MichelBartels committed:
Added dataset to table entries

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

julian-risch commented on docs/latest/guides/model_distillation.mdx:
```diff
@@ -0,0 +1,67 @@
+Here are the results of some of our test and the parameters used:
+| Student model | Teacher model | Dataset | `temperature` | `distillation_loss_weight` | Distilled model EM | Distilled model F1 | Baseline EM | Baseline F1 | Teacher EM | Teacher F1 |
+|-|-|-|-|-|-|-|-|-|-|-|
+| prajjwal1/bert-medium | deepset/bert-large-uncased-whole-word-masking-squad2 | SQuADv2 | 5.0 | 1.0 | 68.6% | 72.8% | 65.6% | 69.6% | 78.9% | 83.2%7 
```
I see that the table already has many columns. However, one thing that's currently missing but important is the name (with a link to the model hub page) of the distilled model. Please add these names and links.

MichelBartels replied:
I have added the links now.

MichelBartels replied:
The first link is not yet working as I am still working on the model upload. I will write an update message later when it's finished.

MichelBartels replied:
All models are uploaded now.

julian-risch commented on docs/latest/guides/model_distillation.mdx:
```diff
@@ -0,0 +1,67 @@
+
+## What performance should I expect?
+
+Here are the results of some of our test and the parameters used:
```
Maybe "some of our test" -> "some of our tests"? Maybe we can stress here something like: "We are actively working on extending model distillation features in haystack. The following table shows experiment results of the first version of model distillation in haystack. Stay tuned for updates!"

MichelBartels replied:
Okay, that first one was a typo. I added the second part.

julian-risch commented on docs/latest/guides/model_distillation.mdx:
```diff
@@ -0,0 +1,67 @@
+# Model distillation
```
Let's capitalize this headline and also the other headlines as in the other guides on the documentation website. Here: "Model Distillation"

MichelBartels replied:
Okay, I have changed that.

julian-risch commented on docs/latest/guides/model_distillation.mdx:
```diff
@@ -0,0 +1,67 @@
+
+## Why should I want to do distil a language model?
+
+In most cases, a larger model size leads to better performance. But the more parameters a model has, the more resource-intensive it is. This makes deployment more difficult and increases latency.
```
Let's replace "performance" with "result quality" to disambiguate here. It's not about runtime, for example.

MichelBartels replied:
I replaced it with quality of results and accuracy. I hope that's also okay.

julian-risch commented on docs/latest/guides/model_distillation.mdx:
```diff
@@ -0,0 +1,67 @@
+## Why should I want to do distil a language model?
+
+In most cases, a larger model size leads to better performance. But the more parameters a model has, the more resource-intensive it is. This makes deployment more difficult and increases latency.
+For this reason, choosing the right reader is always a trade-off between performance and deployment cost. Model distillation is one way to bridge the performance gap between a large language model (the teacher) and a small model (the student).
```
Let's replace "performance" with "result quality" here too.

MichelBartels replied:
Addressed in answer above

julian-risch commented on docs/latest/guides/model_distillation.mdx:
```diff
@@ -0,0 +1,67 @@
+In most experiments, we found a temperature between 1 and 5 to be most useful.
+
+`distillation_loss_weight` specifies the weight given to the distillation loss in relation to the loss based on the label.
+Most times, you should set this to a relatively high value (e.g. 0.75).
```
Please add here two sentences on what the extreme values mean (0 and 1). I think it helps users understanding the parameter, explains that we don't always need labels for model distillation and also thereby mentions the range of this parameter (between 0 and 1).

MichelBartels replied:
Okay, I have added a few sentences explaining that.

julian-risch reviewed (changes requested):
Hi @MichelBartels the text looks already quite good to me. Just left some minor comments also to make sure that this guide is similar to the other ones we already have.
Regarding the table I am still a bit unsure what columns to include or whether to change rows with columns. Maybe we could also split it in two tables, for example, one with F1 score and one with exact match. I am afraid that the table is too large at the moment.

julian-risch commented:
Maybe we could also add a sentence to explain that the current model distillation method that we are using aims to improve the result quality of small models. In contrast, pruning techniques would try to make an existing model smaller by keeping the same result quality. I liked your "bridging the gap" sentence here! 👍 Maybe we can elaborate a bit more there. Users might have heard about model distillation or distilled models somewhere else and it could be that they are expecting something else than what we provide.

brandenchan commented:
Cool to see this in action! Thanks for your work on distillation and also on documenting it. Just want to let you know that to make this page show on the website you will need to add an entry into ` haystack-website/docs/latest/menu.json` - this file defines what will appear on the left Table of Contents.

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels committed:
Update model_distillation.mdx

MichelBartels commented:
I have now added a sentence referring to pruning. I also split the table into two. One for the parameters and one for the results.

julian-risch commented:
Then what's left is mainly editing the `haystack-website/docs/latest/menu.json` file

MichelBartels committed:
Update menu.json

MichelBartels commented:
I have now updated the menu.json. However, I am not entirely certain I did this correctly. I have added an entry smilarly to the other guides, but haven't specified the filename. I am not sure where I have to add that.

MichelBartels commented:
Currently, the guide isn't shown in the sidebar. Also, despite working in the Github preview the table isn't correctly rendered. I am not quite sure what is causing those issues

@brandenchan Could you help with this?

julian-risch commented on docs/latest/menu.json:
```diff
@@ -46,7 +46,8 @@
       {"slug": "chatbots", "title": "Chatbot Integration"},
-      {"slug": "table-qa", "title": "Table Question Answering"}
+      {"slug": "table-qa", "title": "Table Question Answering"},
+      {"slug": "model-distillation", "title": "Model Distillation"},
```
@MichelBartels The last comma here behind `"Model Distillation"}` could be a problem, I think.

MichelBartels replied:
I have now changed this line. This would make sense as trailing commas aren't part of the JSON specification.

MichelBartels committed:
Update menu.json

julian-risch reviewed (approved):
Let's wait with merging until the preview in vercel looks good.

MichelBartels committed:
Recreated tables using tabletomarkdown.com

MichelBartels committed:
making model distillation table smaller

MichelBartels committed:
splitting distillation table into two

MichelBartels committed:
removing EM from model distillation table

MichelBartels committed:
fix table

MichelBartels committed:
add column to first model distillation table
